main/src/ca/uwaterloo/flix/language/phase/MonoTypes.scala

```diff
@@ -308,7 +310,14 @@ object MonoTypes {
         case Type.Cst(_, _) =>
           tpe
         case Type.Apply(tpe1, tpe2, loc) =>
-          Type.Apply(visitType(tpe1), visitType(tpe2), loc)
+          val t1 = visitType(tpe1)
+          val t2 = visitType(tpe2)
+          // Performance: Reuse tpe0, if possible.
+          if ((t1 eq tpe1) && (t2 eq tpe2)) {
+            tpe
+          } else {
+            Type.Apply(t1, t2, loc)
```

Comment:
Should we have a helper for this?

Reply:
Its kind of hard, because its specific. I would wait a bit and see.

Reply:
We could funnel everything through Type.mkApply-- but I would wait just a bit before doing that,